pkg/koordlet/runtimehooks/hooks/terwayqos/terwayqos.go

```diff
@@ -0,0 +1,289 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package terwayqos
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/klog/v2"
+
+	"github.com/koordinator-sh/koordinator/apis/extension"
+	slov1alpha1 "github.com/koordinator-sh/koordinator/apis/slo/v1alpha1"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/resourceexecutor"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/runtimehooks/hooks"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/runtimehooks/rule"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/statesinformer"
+	"github.com/koordinator-sh/koordinator/pkg/koordlet/util"
+)
+
+const (
+	rootPath   = "/var/lib/terway/qos"
+	podConfig  = "pod.json"
+	nodeConfig = "node.json"
+)
+
+const (
+	name        = "TerwayQoS"
+	description = "network qos management"
+
+	ruleNameForNodeQoS = name + " (nodeQoS)"
+	ruleNameForAllPods = name + " (allPods)"
+)
+
+type Plugin struct {
+	executor resourceexecutor.ResourceUpdateExecutor
+
+	lock sync.RWMutex
+	node *Node
+}
+
+func (p *Plugin) Register(op hooks.Options) {
+	klog.V(5).Infof("register hook %v", "terwqy qos configure generator")
+	rule.Register(ruleNameForNodeQoS, description,
+		rule.WithParseFunc(statesinformer.RegisterTypeNodeSLOSpec, p.parseRuleForNodeSLO),
+		rule.WithUpdateCallback(p.update))
+	rule.Register(ruleNameForAllPods, description,
```

Comment:
use hooks.Register(rmconfig.PreRunPodSandbox,) can make sure pod can get network qos before they start.

Reply:
The PreRunPodSandbox does not contain Pod information, so aspects like the Pod QoS level cannot be determined. Additionally, the synchronization of QoS settings with Terway is asynchronous and not guaranteed to take effect before startup. This is not supported at the moment.

Reply:
// TODO can parse qos from cgroup dir